pkg/common/security/usergroup_test.go

```diff
@@ -19,34 +19,79 @@
 package security
 
 import (
+	"fmt"
+	"reflect"
 	"strings"
 	"testing"
+	"time"
 
 	"gotest.tools/v3/assert"
 
+	"github.com/apache/yunikorn-core/pkg/common"
 	"github.com/apache/yunikorn-scheduler-interface/lib/go/si"
 )
 
+func (c *UserGroupCache) getUGsize() int {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return len(c.ugs)
+}
+
+func (c *UserGroupCache) getUGGroupSize(user string) int {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	ug := c.ugs[user]
+	return len(ug.Groups)
+}
+
+func (c *UserGroupCache) getUGmap() map[string]*UserGroup {
+	c.lock.RLock()
+	defer c.lock.RUnlock()
+	return c.ugs
+}
+
 func TestGetUserGroupCache(t *testing.T) {
 	// get the cache with the test resolver set
 	testCache := GetUserGroupCache("test")
-	if testCache == nil {
-		t.Fatal("Cache create failed")
-	}
-	if len(testCache.ugs) != 0 {
-		t.Errorf("Cache not empty: %v", testCache.ugs)
-	}
+	assert.Assert(t, testCache != nil, "Cache create failed")
+	assert.Equal(t, 0, testCache.getUGsize(), "Cache is not empty: %v", testCache.getUGmap())
+
+	testCache.Stop()
+	assert.Assert(t, instance == nil, "instance should be nil")
+	assert.Assert(t, stopped.Load())
+
+	// get the cache with the os resolver set
+	testCache = GetUserGroupCache("os")
+	assert.Assert(t, testCache != nil, "Cache create failed")
+	assert.Equal(t, 0, testCache.getUGsize(), "Cache is not empty: %v", testCache.getUGmap())
+
+	testCache.Stop()
+	assert.Assert(t, instance == nil, "instance should be nil")
+	assert.Assert(t, stopped.Load())
+
+	// get the cache with the default resolver set
+	testCache = GetUserGroupCache("unknown")
+	assert.Assert(t, testCache != nil, "Cache create failed")
+	assert.Equal(t, 0, testCache.getUGsize(), "Cache is not empty: %v", testCache.getUGmap())
+
+	testCache.Stop()
+	assert.Assert(t, instance == nil, "instance should be nil")
+	assert.Assert(t, stopped.Load())
+
+	// test for re stop again
+	testCache.Stop()
+	assert.Assert(t, instance == nil, "instance should be nil")
+	assert.Assert(t, stopped.Load())
 }
 
 func TestGetUserGroup(t *testing.T) {
 	testCache := GetUserGroupCache("test")
 	testCache.resetCache()
 	// test cache should be empty now
-	if len(testCache.ugs) != 0 {
-		t.Fatalf("Cache not empty: %v", testCache.ugs)
-	}
+	assert.Equal(t, 0, testCache.getUGsize(), "Cache is not empty: %v", testCache.getUGmap())
 
 	ug, err := testCache.GetUserGroup("testuser1")
+	testCache.lock.Lock()
```

Comment:
This lock is not needed here. I believe it should be moved right before L#106 when you directly access `testCache.ugs`. `GetUserGroup()` obtains a lock to read the map and returns a **copy** of the `UserGroup` type, not a pointer.